File: languages/kotlin/k-means.kt
// k-means: Lloyd's clustering algorithm - the machine-learning axis of the suite. Cluster N
// integer D-dimensional points into K clusters over ITERS fixed iterations: assign each point
// to its nearest centroid (integer squared Euclidean distance), then recompute each centroid as
// the floor-mean of its members. Everything is integer (quantized-style) - deterministic, no
// floating point, so no FMA / summation-order divergence across languages.
//
// Pinned tie-breaks: a point ties to the LOWEST-index centroid (strict < while scanning); an
// empty cluster keeps its centroid unchanged. The checksum hashes the final centroids and the
// final assignment of every point.

const val P = 1000000007L
const val K = 16
const val D = 4
const val ITERS = 10
const val RANGE = 256L

fun kMeans(n: Int): Long {
    val pt = LongArray(n * D)
    var s = 42L
    for (i in 0 until n * D) {
        s = (s * 1103515245L + 12345L) and 0x7fffffffL
        pt[i] = s % RANGE
    }
    val cen = LongArray(K * D)
    for (i in 0 until K * D) cen[i] = pt[i]            // initial centroids = first K points
    val assign = IntArray(n)

    for (iter in 0 until ITERS) {
        for (i in 0 until n) {                         // assignment
            var best = 0
            var bd = -1L
            for (k in 0 until K) {
                var dist = 0L
                for (d in 0 until D) {
                    val df = pt[i * D + d] - cen[k * D + d]
                    dist += df * df
                }
                if (bd < 0 || dist < bd) { bd = dist; best = k }
            }
            assign[i] = best
        }
        val ssum = LongArray(K * D)                    // update: floor-mean, empty unchanged
        val cnt = LongArray(K)
        for (i in 0 until n) {
            val k = assign[i]
            cnt[k]++
            for (d in 0 until D) ssum[k * D + d] += pt[i * D + d]
        }
        for (k in 0 until K) {
            if (cnt[k] > 0) {
                for (d in 0 until D) cen[k * D + d] = ssum[k * D + d] / cnt[k]
            }
        }
    }

    for (i in 0 until n) {                             // final assignment with final centroids
        var best = 0
        var bd = -1L
        for (k in 0 until K) {
            var dist = 0L
            for (d in 0 until D) {
                val df = pt[i * D + d] - cen[k * D + d]
                dist += df * df
            }
            if (bd < 0 || dist < bd) { bd = dist; best = k }
        }
        assign[i] = best
    }

    var h = 0L
    for (i in 0 until K * D) h = (h * 31 + cen[i]) % P
    for (i in 0 until n) h = (h * 31 + assign[i]) % P
    return h
}

fun main(args: Array<String>) {
    val n = if (args.isNotEmpty()) args[0].toInt() else 8000
    println(kMeans(n))
    println("k-means($n)")
}

File: languages/c/k-means.c
// k-means: Lloyd's clustering algorithm - the machine-learning axis of the suite. Cluster N
// integer D-dimensional points into K clusters over ITERS fixed iterations: assign each point
// to its nearest centroid (integer squared Euclidean distance), then recompute each centroid as
// the floor-mean of its members. Everything is integer (quantized-style) - deterministic, no
// floating point, so no FMA / summation-order divergence across languages.
//
// Pinned tie-breaks: a point ties to the LOWEST-index centroid (strict < while scanning); an
// empty cluster keeps its centroid unchanged. The checksum hashes the final centroids and the
// final assignment of every point.
#include <stdio.h>
#include <stdlib.h>

#define P 1000000007L
#define K 16
#define D 4
#define ITERS 10
#define RANGE 256

static long lcg(long s) { return (s * 1103515245L + 12345L) & 0x7fffffffL; }

int main(int argc, char **argv) {
    int N = argc > 1 ? atoi(argv[1]) : 8000;
    long *pt = malloc((size_t)N * D * sizeof(long));
    long s = 42;
    for (long i = 0; i < (long)N * D; i++) { s = lcg(s); pt[i] = s % RANGE; }
    long cen[K * D];
    for (int i = 0; i < K * D; i++) cen[i] = pt[i];   // initial centroids = first K points
    int *assign = malloc((size_t)N * sizeof(int));

    for (int iter = 0; iter < ITERS; iter++) {
        for (int i = 0; i < N; i++) {                 // assignment
            int best = 0; long bd = -1;
            for (int k = 0; k < K; k++) {
                long dist = 0;
                for (int d = 0; d < D; d++) { long df = pt[(long)i * D + d] - cen[k * D + d]; dist += df * df; }
                if (bd < 0 || dist < bd) { bd = dist; best = k; }
            }
            assign[i] = best;
        }
        long ssum[K * D] = {0}; long cnt[K] = {0};    // update: floor-mean, empty unchanged
        for (int i = 0; i < N; i++) {
            int k = assign[i]; cnt[k]++;
            for (int d = 0; d < D; d++) ssum[k * D + d] += pt[(long)i * D + d];
        }
        for (int k = 0; k < K; k++)
            if (cnt[k] > 0)
                for (int d = 0; d < D; d++) cen[k * D + d] = ssum[k * D + d] / cnt[k];
    }

    for (int i = 0; i < N; i++) {                      // final assignment with final centroids
        int best = 0; long bd = -1;
        for (int k = 0; k < K; k++) {
            long dist = 0;
            for (int d = 0; d < D; d++) { long df = pt[(long)i * D + d] - cen[k * D + d]; dist += df * df; }
            if (bd < 0 || dist < bd) { bd = dist; best = k; }
        }
        assign[i] = best;
    }

    long h = 0;
    for (int i = 0; i < K * D; i++) h = (h * 31 + cen[i]) % P;
    for (int i = 0; i < N; i++) h = (h * 31 + assign[i]) % P;
    printf("%ld\n", h);
    printf("k-means(%d)\n", N);
    return 0;
}
